std/numeric.d

```diff
@@ -3329,6 +3329,118 @@ C swapRealImag(C)(C input)
     return C(input.im, input.re);
 }
 
+/**
+Finds the mean (colloquially known as the average) of a range. If `r` does
+not provide the `length` member, then this function will do element by
+element summation, rather than the more accurate methods provided by
+$(REF sum, std,algorithm,iteration).
+
+An optional parameter `seed` is used for non-builtin types in order to
+correctly seed the summation with the equivalent to `0`.
+
+This function will return `real.nan` if the range is empty and if the element
+type of `r` is a built-in numerical type. However, if the seed value is defined,
+then empty ranges will return `ElementType!R.init`.
+
+This function is $(BIGOH r.length).
+
+Params:
+    r = An $(REF_ALTTEXT input range, isInputRange, std,range,primitives)
+    seed = An optional parameter for taking the mean of types which can't
+    implicitly convert to `real`. This value will initialize the summation,
+    so it should be equivalent to `0`.
+
+Returns:
+    the mean of r
+*/
+real mean(R)(R r)
+    if (isInputRange!R &&
+        is(ElementType!R : real) &&
+        !isInfinite!R)
+{
+    if (r.empty)
+    {
+        return real.nan;
+    }
+
+    return mean(r, real(0.0));
```

Comment:
> I assume that if you want statistical values, accuracy will be more valuable than speed.

FWIW with the test provided below `double` yields exactly the same inaccuracy (`25002500002500001792.000000`).
The point was that always picking `real` is a severe performance penalty with almost no neglectable accuracy improvements, so the user should at least have the choice.
Also there's another, very valid point about allowing `double`: it's consistent on all platforms. I purposefully ignored `real` on my Tinflex random project last year because the algorithm was very sensitive to small change ([more details about FP behavior with dmd](http://forum.dlang.org/post/hjaiavlfkoamenidomsa@forum.dlang.org) or as a [handy list of FP issues to keep in mind](https://github.com/libmir/mir/wiki/Floating-point-issues)).